pkg/pillar/cmd/zedrouter/networkinstance.go

```diff
@@ -1870,3 +1860,178 @@ func uplinkToPhysdev(ifname string) string {
+	status *types.NetworkInstanceStatus) bool {
+	if !isValidIPRange(status.DhcpRange) {
+		return isValidIPSubnet(status.Subnet)
+	}
+	return false
+}
+
+func validateIPConfig(ctx *zedrouterContext,
+	status *types.NetworkInstanceStatus) {
+	status.IsValidIPRange = isValidIPRange(status.DhcpRange)
+	status.IsValidIPSubnet = isValidIPSubnet(status.Subnet)
+	status.IsValidIPConfig = status.IsValidIPSubnet || status.IsValidIPRange
+	publishNetworkInstanceStatus(ctx, status)
+}
+
+// initialize the Host IP Address Management Module
+func handleNetworkInstanceIPAddrInit(ctx *zedrouterContext,
+	status *types.NetworkInstanceStatus) {
+
+	validateIPConfig(ctx, status)
+	if !status.IsValidIPConfig {
+		return
+	}
+	addrcount := 253
+	// get host IP Address count:
+	// if DhcpRange is set, derive from it
+	// otherwise derive from the subnet
+	if status.IsValidIPRange {
+		addrcount = getIPRangeAddrCount(status.DhcpRange)
+	} else {
+		if !status.IsValidIPSubnet {
+			return
+		}
+		addrcount = getIPSubnetAddrCount(status.Subnet)
+	}
+	// XXX: upper limit
+	if addrcount > 256 {
+		addrcount = 253
+	}
+	status.HostIPAdressCount = addrcount
+	log.Functionf("Initialized Network instance %s with addressable count %d\n",
+		status.DisplayName, addrcount)
+	publishNetworkInstanceStatus(ctx, status)
+}
+
+// for an App Inastance attached to an underlay network,
+// based on the appNum, try to get a free index within a
+// assignable ip address range
+func appNumOnUNetAllocate(ctx *zedrouterContext,
+	status *types.NetworkInstanceStatus, appNum int) int {
+	if !status.IsValidIPConfig {
+		return 0
+	}
+	// take the mod value for the range, on AppNum
+	appNumMod := appNum % status.HostIPAdressCount
+	appNumOnUNet := appNumMod
+
+	// if collision, find the next free bit
+	if status.AppNumBitMap.IsSet(appNumOnUNet) {
+		log.Functionf("AppNumOnUNetAllocate: Hash Collision detected for appNum %d on Network %s at %d\n",
+			appNum, status.DisplayName, appNumMod)
+		i := appNumMod + 1
+		// take next free bit value
+		for i = appNumMod + 1; i <= status.HostIPAdressCount; i++ {
+			if !status.AppNumBitMap.IsSet(i) {
+				appNumOnUNet = i
+				break
+			}
+		}
+		// reached the max count
+		if i > status.HostIPAdressCount {
+			// start at the first bit value and iterate
+			// for next free bit
+			for i := 1; i < appNumMod; i++ {
+				if !status.AppNumBitMap.IsSet(i) {
+					appNumOnUNet = i
+					break
+				}
+			}
+		}
+		// could not find a free slot
+		if i == appNumMod {
+			log.Fatalf("appNumOnUNetAllocate: on network %s failed for %d\n", status.DisplayName, appNum)
+		}
+	}
+	// set the bit value, and publish
+	status.AppNumBitMap.Set(appNumOnUNet)
+	log.Functionf("Allocating appNum %d on Network %s\n",
+		appNumOnUNet, status.DisplayName)
+	publishNetworkInstanceStatus(ctx, status)
+	return appNumOnUNet
+}
+
+func appNumOnUNetFree(ctx *zedrouterContext,
+	status *types.NetworkInstanceStatus, appNumOnUNet int) {
+	if !status.IsValidIPConfig {
+		return
```

Comment:
Why is this check needed?
The check which is needed (here or in the caller) is that the allocation was actually done.
Maybe that is appNumOnUNet != 0?

Reply:
Have simplified the code. This functions generate/clear the appNumOnUNet, hence check is against the HostIPAddressCount.